sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobContainerAsyncClient.java

```diff
@@ -1045,4 +1045,47 @@ private boolean validateNoETag(BlobRequestConditions modifiedRequestConditions)
         }
         return modifiedRequestConditions.getIfMatch() == null && modifiedRequestConditions.getIfNoneMatch() == null;
     }
+
+    /**
+     * Restores a previously deleted container.  The restored container
+     * will be renamed to the name of this {@link BlobContainerAsyncClient}.
```

Comment:
At a higher level, is this what the user would expect? My naive expectation would be that the container is restored into its previous container name?

Reply:
Yes. That's how the underlying API works. However there's nothing stopping customer from restoring into same name if nobody created container with same name.

Reply:
That's not quite my question - I'm not asking how the API works, I was seeking to clarify that this aligns with the users expectations (in the absence of any particular API or implementation).

Reply:
Oh I think I see what you mean. `restore` sounds like it should be acting on current `BlobContainerClient` so current implementation may sound little bit backwards. What we have here now is more of "create from deleted container". 
I think we have two options now. Either change wording here or reverse this API.

Changing wording would be:
`client.createFrom(string deletedContainerName, string deleteContainerVersion)` or something like that.

Reversing would mean two overloads:
`client.restore(string deletedVersion)` - this would preserve name
`client.restore(string deletedVersion, string destinationContainerName)` - that would land in renamed container.

@JonathanGiles  what do you think?

Also looping in @seanmcc-msft @tg-msft for this discussion. I guess we'd want to be consistent in .NET if we are changing that (PR reference https://github.com/Azure/azure-sdk-for-net/pull/11244 ). And @xiafu-msft  for Python implementation.

Reply:
The current .NET implementation is on the BlobContainerClient:

```
        /// <summary>
        /// Restores a previously created container.  The restored container
        /// will be renamed to the name of this <see cref="BlobContainerClient"/>.
        /// If the container associated with this <see cref="BlobContainerClient"/>
        /// already exists, this call will result in a 409 (conflict).
        /// This API is only functional is Container Soft Delete is enabled
        /// for the storage account associated with the container.
        /// </summary>
               public virtual Response Restore(
            string deletedContainerName,
            string deletedContainerVersion,
            CancellationToken cancellationToken = default) {...}
```

The current user experience for .NET:
```
BlobContainerClient container = 
    blobServiceClient.GetBlobContainerClient("nameToRestoreContainerTo";
await container.RestoreAsync("deletedContainerName", "deletedContainerVersion");
```
The restored container will "become" the container associated with this BlobContainerClient, and now users can interact with it via this client.

For the 3 proposed options:
`client.createFrom(string deletedContainerName, string deleteContainerVersion) `
I'm not a huge fan of this one.  I'd rather stick with "Restore" in the API name to align with the service.  Also, the name "createFrom"could be confusing to new users who are just trying to create a Container.

`client.restore(string deletedVersion)`
I could get behind this one.

`client.restore(string deletedVersion, string destinationContainerName)`
I think this one doesn't align super well with our current user experience.  Currently, a BlobContainerClient only interact with the Container it is associated with.  This would break that convention.

`BlobServiceClient.restoreContainer(string deletedContainerName, string deletedContainerVersion)`
`BlobServiceClient.restoreContainer(string deletedContainerName, string deletedContainerVersion, string destinationContainerName)`
These are ok, I'd want `restoreContainer` to return a BlobContainerClient for convenience.

I prefer the current .NET implementation the most.

Reply:
I'd recommend storage discuss this internally and work with @tg-msft to land that across the languages as necessary.

Reply:
I think I agree with Sean and what Kamil has. Best option would be to align with undelete on blobs, but since that only restores in place, it doesn't offer much guidance here. Perhaps the next best thing is to look at copy, which has the client target the destination and then passes the source as a parameter. I think what .NET is doing most closely aligns with that, and if undelete for blobs ever allows you to move the blob as a part of that operation, we should have the source be a parameter.

I definitely think there should be an easy restore-in-place overload for cases where the source and destination are the same, so the customer just doesn't need to pass in a name at all. And I think we should rework the javadocs here to be a little clearer about what's happening. 

Also, for the sake of consistency with blobs, I think we should call this undelete. Since we link to the REST docs in our javadocs, I don't really see a likelihood that customers will get lost hunting through rest docs for the corresponding method.

Reply:
In terms of existing APIs and patterns, this feels like a mix of [Blob's Undelete](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs#L987) and [DataLake's Rename](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Files.DataLake/api/Azure.Storage.Files.DataLake.netstandard2.0.cs#L45) operation.

I really like Sean's suggestion because I think that removes all the confusion about how to use the API:

```C#
class BlobServiceClient
{
    Response<BlobContainerClient> RestoreContainer(
        string deletedContainerName,
        string deletedContainerVersion,
        string destinationContainerName = null);
}
```

This doesn't hide the `destinationContainerName` concept in the BlobContainerClient you use to invoke the API.  Instead it's something you can pass if you want it to change or leave null to keep the same name.

Reply:
@tg-msft Both undelete and rename have the client pointing to the target rather than accessing them from the parent container. If restore is similar to both of those, why would introduce a different calling pattern?

Reply:
By looking at `BlobContainerClient.create` and `BlobContainerClient.delete` family (which seem to be relatives of restore - you have to create and delete container in order to restore, same story for .NET) I noticed that we have them both at `BlobContainerClient` and `BlobServiceClient`. So, I'd suggest to keep what we already have at Container level (plus extra overload that doesn't take name) and add mirroring methods to ServiceClient.

@tg-msft @seanmcc-msft what do you think?

Reply:
@kasobol-msft, sounds good.

So for STG 73, it looks like users will only be able to restore containers under their original name (based on Emily's ongoing email thread with them, there's a bug), with the restore + rename functionality coming in 75.  So we could hold off on the overload that takes new name for now.

Reply:
I completely agree on `Undelete` vs `Restore`.  We should use the same name for everything and I've left that feedback on one of the .NET PRs.  In .NET it doesn't matter which way we move thanks to `[EditorBrowsable(Never)]`, but I'd vote for consistency with whatever the other languages are going to call this.

The reason I like moving this method up to the parent client is because the return type makes more sense.  When I call `BlobClientContainer.Rename(name)`, I get why you're returning a new client _(if I understand that clients are effectively cloud pointers)_.  If I call `BlobContainerClient.Restore()` I have no idea why it's returning a new client instance _("is it actually the same instance, or is it a new instance slightly different from my last one, or ...?")_.  If I call `BlobServiceClient.Restore(name)` it's really easy to understand that the returned client represents the restored container.  I don't feel strongly about this if everyone else prefers keeping it on `BlobContainerClient`.

The one thing I wouldn't do is put it on both `BlobServiceClient` and `BlobContainerClient`.  That pattern is an on-ramp where we have GetClient/List/Meta-operations on the parent, Create/Delete on both, and everything else on the child.  If you want to explore expanding the set of operations on both, we should push the architects to clarify the guidelines around hierarchical clients because they're not all written down today.  That discussion was a very large can of worms.  :smile:

Reply:
Thanks @tg-msft @seanmcc-msft @rickle-msft for inputs.

I'm going to pull that API up to service client.

The prevailing reason for me would be that we actually need at least a tuple - name and versionid of deleted container - something that's only available at `BlobServiceClient` level through listing. So one won't be able to complete the whole experience with `BlobContainerClient` alone.

Reply:
https://github.com/Azure/azure-sdk-for-net/pull/12025